ext/standard/streamsfuncs.c

```diff
@@ -616,97 +616,150 @@ PHP_FUNCTION(stream_get_wrappers)
+#define FD_BIGSET_ISSET(fd, set) ((set)->fds_bits[(fd) / 8] & (1 << ((fd) % 8)))
+#define FD_BIGSET_CLR(fd, set) ((set)->fds_bits[(fd) / 8] &= ~(1 << ((fd) % 8)))
+#define FD_BIGSET_FREE(set) do { \
+    if ((set)->fds_bits) { \
+        efree((set)->fds_bits); \
+        (set)->fds_bits = NULL; \
+    } \
+    (set)->size = 0; \
+} while (0)
+#else
+#define FD_BIGSET_ZERO(set, num_fds) do { \
+    (set)->size = (num_fds + 7) / 8; \
+    (set)->__fds_bits = (char *)ecalloc((set)->size, sizeof(char)); \
+} while (0)
+#define FD_BIGSET_SET(fd, set) ((set)->__fds_bits[(fd) / 8] |= (1 << ((fd) % 8)))
+#define FD_BIGSET_ISSET(fd, set) ((set)->__fds_bits[(fd) / 8] & (1 << ((fd) % 8)))
+#define FD_BIGSET_CLR(fd, set) ((set)->__fds_bits[(fd) / 8] &= ~(1 << ((fd) % 8)))
+#define FD_BIGSET_FREE(set) do { \
+    if ((set)->__fds_bits) { \
+        efree((set)->__fds_bits); \
+        (set)->__fds_bits = NULL; \
+    } \
+    (set)->size = 0; \
+} while (0)
+#endif
 
-	if (Z_TYPE_P(stream_array) != IS_ARRAY) {
-		return 0;
-	}
+int stream_array_to_fd_bigset(zval *stream_array, fd_bigset *set, php_socket_t *max_fd) {
+    zval *elem;
+    php_stream *stream;
+    int cnt = 0;
 
-	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(stream_array), elem) {
-		/* Temporary int fd is needed for the STREAM data type on windows, passing this_fd directly to php_stream_cast()
-			would eventually bring a wrong result on x64. php_stream_cast() casts to int internally, and this will leave
-			the higher bits of a SOCKET variable uninitialized on systems with little endian. */
-		php_socket_t this_fd;
+    if (Z_TYPE_P(stream_array) != IS_ARRAY) {
+        return 0;
+    }
 
-		ZVAL_DEREF(elem);
-		php_stream_from_zval_no_verify(stream, elem);
-		if (stream == NULL) {
-			continue;
-		}
-		/* get the fd.
-		 * NB: Most other code will NOT use the PHP_STREAM_CAST_INTERNAL flag
-		 * when casting.  It is only used here so that the buffered data warning
-		 * is not displayed.
-		 * */
-		if (SUCCESS == php_stream_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT | PHP_STREAM_CAST_INTERNAL, (void*)&this_fd, 1) && this_fd != -1) {
+    ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(stream_array), elem) {
+        php_socket_t this_fd;
 
-			PHP_SAFE_FD_SET(this_fd, fds);
+        ZVAL_DEREF(elem);
+        php_stream_from_zval_no_verify(stream, elem);
+        if (stream == NULL) {
+            continue;
+        }
 
-			if (this_fd > *max_fd) {
-				*max_fd = this_fd;
-			}
-			cnt++;
-		}
-	} ZEND_HASH_FOREACH_END();
-	return cnt ? 1 : 0;
-}
+        if (SUCCESS == php_stream_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT | PHP_STREAM_CAST_INTERNAL, (void*)&this_fd, 1) && this_fd != -1) {
+            FD_BIGSET_SET(this_fd, set);
 
-static int stream_array_from_fd_set(zval *stream_array, fd_set *fds)
-{
-	zval *elem, *dest_elem;
-	HashTable *ht;
-	php_stream *stream;
-	int ret = 0;
-	zend_string *key;
-	zend_ulong num_ind;
+            if (this_fd > *max_fd) {
+                *max_fd = this_fd;
+            }
```

Comment:
Won't this break coherency if this code is hit? If max_fd changes here, you possibly already crashed in `FD_BIGSET_SET` due to a segfault. If you made it this far, it should probably be treated as an error; and probably be detected before `FD_BIGSET_SET` to prevent a segfault.

Reply:
Not sure what you mean. That code was already in there from before my commit.